core/src/databases/table.rs

```diff
@@ -383,46 +394,140 @@ impl LocalTable {
+
+        // Tell the worker that there are things to process for this table.
+        let upsert_call = TableUpsertActivityData {
+            time: utils::now(),
+            project_id: self.table.project().project_id(),
+            data_source_id: self.table.data_source_id().to_string(),
+            table_id: self.table.table_id().to_string(),
+        };
+        let _: () = redis_conn
+            .hset(
+                REDIS_TABLE_UPSERT_HASH_NAME,
+                self.table.unique_id(),
+                serde_json::to_string(&upsert_call)?,
+            )
+            .await?;
+
+        Ok(())
+    }
+
     pub async fn upsert_rows(
         &self,
         store: Box<dyn Store + Sync + Send>,
         databases_store: Box<dyn DatabasesStore + Sync + Send>,
         rows: Vec<Row>,
         truncate: bool,
     ) -> Result<()> {
-        let rows = Arc::new(rows);
+        Self::validate_rows_lowercase(rows.clone()).await?;
+
+        if truncate {
+            let now = utils::now();
+            let lock_manager = LockManager::new(vec![REDIS_URI.clone()]);
+
+            // Take the table lock to make sure we don't conflict with processing
+            // happening in the background worker.
+            let lock = lock_manager
+                .acquire_no_guard(
+                    self.table.get_background_processing_lock_name().as_bytes(),
+                    Duration::from_secs(REDIS_LOCK_TTL_SECONDS),
+                )
+                .await?;
 
-        let mut now = utils::now();
-        // Validate that all rows keys are lowercase. We run it in a spawn_blocking since it is CPU
-        // bound (even if running fast for resaonably sized tables);
-        {
-            let rows = rows.clone();
-            tokio::task::spawn_blocking(move || {
-                for (row_index, row) in rows.iter().enumerate() {
-                    match row.value().keys().find(|key| match key.chars().next() {
-                        Some(c) => c.is_ascii_uppercase(),
-                        None => false,
-                    }) {
-                        Some(key) => Err(anyhow!(
-                            "Row {} has a key '{}' that contains uppercase characters",
-                            row_index,
-                            key
-                        ))?,
-                        None => (),
-                    }
-                }
-                Ok::<_, anyhow::Error>(())
-            })
-            .await??;
+            info!(
+                lock_acquisition_duration = utils::now() - now,
+                "Upsert lock acquired in upsert_rows"
+            );
+
+            // Since truncate replaces everything, we get rid of all non-truncate and row deletion
+            // pending operations that got queued before we got called.
+            // And those that arrive later cannot happen until we release the lock.
+            GoogleCloudStorageBackgroundProcessingStore::delete_all_files_for_table(&self.table)
```

Comment:
I still don't quite get how this can be correct on the GCS side if we delete on truncate vs pushing the truncate command down the async pipeline?

As discused if we have the sequence write rows r_A, write rows r_B, truncate + write rows r_C we can end up with r_A r_B being written on top of r_C no?